### PR TITLE
Add standard `ipo` variant for CMakePackage

### LINKED ISSUE #18373: Standardized CMake Interprocedural Optimization Variant
Add standard `ipo` variant to CMakePackage.

### Rationale
CMake exposes the `CMAKE_INTERPROCEDURAL_OPTIMIZATION` variable to enable IPO/LTO on all targets in a project. Similar to the standard `build_type` variant for CMake, Spack should have a standard `ipo` variant (defaulting to `False`) that enables `CMAKE_INTERPROCEDURAL_OPTIMIZATION`.

### Description
```python
# lib/spack/spack/build_systems/cmake.py
variant('ipo', default=False, description='CMake interprocedural optimization')

def _std_args(pkg):
    ...
    try:
        ipo = pkg.spec.variants['ipo'].value
    except KeyError:
        ipo = False

    args = [
        '-G', generator,
        define('CMAKE_INSTALL_PREFIX', pkg.prefix),
        define('CMAKE_BUILD_TYPE', build_type),
        define_from_variant('CMAKE_INTERPROCEDURAL_OPTIMIZATION', ipo),
    ]
```

### Additional information
Spack 0.15.4-710-125191931

### General information
- [x] I have run `spack --version` and reported the version of Spack
- [x] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by omor1**
Due to CMake's handling of compilers, linkers, archivers, etc., it can be harder to enable IPO/LTO by simply manipulating `CFLAGS`, `AR`, `RANLIB`, etc. than in autotools-based projects. The variable `CMAKE_INTERPROCEDURAL_OPTIMIZATION` has existed since CMake 3.9 to allow compilation with IPO/LTO more easily.

This PR adds a standard `ipo` variant to CMake packages, defaulting to `False`, that allows users to control this variable. Enabling it possibly breaks some packages, but it defaults to off, so I don't think it should cause any problems.

Closes #18373.

### CODE REVIEW

**Review by adamjstewart — APPROVED** [review]
I'll wait for a couple other maintainers to review.

**Comment by adamjstewart** [comment]
> The variable CMAKE_INTERPROCEDURAL_OPTIMIZATION has existed since CMake 3.9 to allow compilation with IPO/LTO more easily.

For older versions of CMake, will the flag be ignored, or could it cause the build to crash? We could also check the version:
```python
if self.spec.satisfies('^cmake@3.9:'):
    args.append(define('CMAKE_INTERPROCEDURAL_OPTIMIZATION', ipo))
```

**Comment by omor1** [comment]
> For older versions of CMake, will the flag be ignored, or could it cause the build to crash? We could also check the version:
> 
> ```python
> if self.spec.satisfies('^cmake@3.9:'):
>     args.append(define('CMAKE_INTERPROCEDURAL_OPTIMIZATION', ipo))
> ```

I think it's just ignored, but it might spit out a warning that confuses users, so that's a good idea. We probably don't want to just ignore the user's variant if they have an incompatible version of CMake, too; let's add a conflict with CMake < 3.9.

File: lib/spack/spack/build_systems/cmake.py
Comment (adamjstewart) [inline lib/spack/spack/build_systems/cmake.py]:
```suggestion
    conflicts('+ipo', when='^cmake@:3.8',
```
Just in case a 3.8.3 patch release comes out someday

Comment (omor1) [reply lib/spack/spack/build_systems/cmake.py]: Does Spack consider 3.8 to consist of everything from 3.8.0 to 3.8.99999... (i.e. 3.8.x)? I saw elsewhere in the documentation suggesting using a .999 release as the endpoint, on the assumption that it will never reach that—but that documentation could be out of date.

For reference, 3.8.2 was released in 2017—I don't think Kitware does LTS/maintenance releases, focusing on the top-of-tree instead. But I do agree with the general point :).

Comment (adamjstewart) [reply lib/spack/spack/build_systems/cmake.py]: Honestly, our version ranges are pretty broken. The 999 stuff is because Spack doesn't support <, only <=. 3.8 should be treated as 3.8.X in most version comparisons.

**Review by junghans — APPROVED** [review]

**Comment by adamjstewart** [comment]
Ping @alalazo

**Review by junghans — APPROVED** [review]

**Review by alalazo — APPROVED** [review]

File: lib/spack/spack/build_systems/cmake.py
Comment (alalazo) [inline lib/spack/spack/build_systems/cmake.py]: Minor suggestion to shorten the message:
```suggestion
              msg='+ipo is not supported by CMake < 3.9')
```

**Comment by alalazo** [comment]
Thanks @omor1 !

**Pull request merged**
